### ISSUE TITLE
improvements to createU

### ISSUE THREAD

**Issue by katzfuss** [katzfuss-group/GPvecchia#27]
createU() should be able to handle covariance matrices or user-supplied covariance functions. Specifically, the argument covmodel should be able to handle pre-specified character strings (currently 'matern' or 'esqe'), an n x (m+1) matrix of covariances corresponding to revNNarray, or a user-specified R function. If a matrix of covariances or an R function is supplied, the argument covparms is ignored.

The R function must have the arguments locs1 and locs2, in that order, where locs1 and locs2 are matrices of locations, with each row corresponding to one location. If this is how the covariance is specified, we can simply compute the n x (m+1) matrix of covariances from the function, and then use the U_NZentries function that can handle the matrix.

Then, we need which way of calling createU() is faster. Test with say, n=40,000 and m=10 in 2D with exponential covariance function. Test if the R covariance function is faster than the C++ computation in U_NZentries with 'matern' and smoothness=0.5.

If the C++ implementation is faster, we should also ensure that the Bessel evaluations are avoided if smoothness=1.5, see e.g. here [https://en.wikipedia.org/wiki/Mat%C3%A9rn_covariance_function#Simplification_for_%CE%BD_half_integer](url)

**Comment by katzfuss** [comment]
Actually, I think the Vecchia approximation, in general, requires a matrix of (m+1) x (m+1) covariances for each location, meaning roughly nm^2 covariances in total. Only for the MRA and some other special cases, we can compute U just based on providing the n x (m+1) covariances mentioned above, using a Cholesky decomposition of a sparsified covariance matrix.

**Comment by marcinjurek** [comment]
Do we want to worry about the general case though? From my (admittedly, limited) perspective, it seems like we only need to pass a matrix when we do filtering. In purely spatial cases covariance will be provided as a parametric family anyways, no? Who would want to do something else? 

Anyways, maybe at least I could start with the MRA case instead of the fully general one?

**Comment by marcinjurek** [comment]
Hello Dr. Katzfuss, so following my understanding of the overleaf document here is what I'm currently doing do assemble the U matrix:

First, I'm assembling a sparse matrix Sigma \kron S, asstated in section 4.3. Let's call this matrix M. Then I calculate L = rev(chol(rev(M))^{-1}). This gives me a sparse lower triangular matrix with the same pattern as the lower triangle of the matrix M.

Since each row of L has at most m+1 nonzero elements I convert L, which is n x n, into an n x (m+1) matrix (let's call it G) by removing zeros from under the diagonal in each row. This leaves me with some zeros in some rows. However, if L[i,j] != 0 then for all k<j L[i,k] != 0 either (in other words, each row consists of two contiguous blocks of nonzeros followed by a block of zeros (this second block might be empty)). I hope it makes sense.

The matrix G described above should be, according to my understanding, the same as the `U.entries$Lentries` matrix calculated by the previous version of the code. Unfortunately, it is not. It is not way off, but it's clearly not identical. One source of discrepancy could be the numerical errors brought about by cholesky and inversion procedures used in either version of the algorithm. But the difference I'm seeing seems a little to big for that. Could you help me figure out what I'm doing wrong?

Here are some illustrations. I use `n=16` points. Then I use 
```{R}
V = vecchia_specify(locs, m=4, conditioning='mra').
```
When I use the old (Rcpp) version of the algorithm here is the U.entries$Lentries matrix I get:
```{R}
               [,1]          [,2]         [,3]     [,4]
 [1,]  1.000000e+00  0.000000e+00  0.000000000 0.000000
 [2,] -8.968786e-03  1.000040e+00  0.000000000 0.000000
 [3,] -5.793755e-04  1.000000e+00  0.000000000 0.000000
 [4,] -1.192346e-03 -8.958099e-03  1.000040930 0.000000
 [5,] -8.020204e-05 -2.571297e-05  1.000000004 0.000000
 [6,] -1.252764e-03 -3.569602e-02  1.000637710 0.000000
 [7,] -8.968762e-03 -4.020548e-05  1.000040220 0.000000
 [8,] -8.957740e-03 -1.192163e-03 -0.008968401 1.000081
 [9,] -3.540088e-02 -1.252125e-03 -0.035401556 1.001264
[10,] -3.538154e-02 -3.510679e-02 -0.035381543 1.001890
[11,] -8.944442e-03 -3.541934e-02 -0.007699830 1.000707
[12,] -2.175757e-04 -3.539963e-02 -0.035401320 1.001264
[13,] -5.320338e-04 -2.548562e-04 -0.035693759 1.000637
[14,] -3.567097e-02 -7.696567e-03 -0.035400515 1.001303
[15,] -3.540158e-02 -4.347626e-06 -0.035401578 1.001264
[16,] -5.765207e-04 -3.931483e-05 -0.035696674 1.000637
```

Contrast it with the G matrix obtained using the algorithm I'm working on:
```{R}
               [,1]          [,2]         [,3] [,4]
 [1,]  1.002687e+00  0.000000e+00  0.000000000    0
 [2,] -7.704147e-03  1.001297e+00  0.000000000    0
 [3,]  1.045684e-04  1.001294e+00  0.000000000    0
 [4,]  9.112194e-05 -6.435360e-03  1.001914429    0
 [5,]  2.085562e-05 -1.343469e-05  1.000677216    0
 [6,]  7.708961e-05 -3.512699e-02  1.000677216    0
 [7,] -6.456851e-03  9.465734e-07  1.001275063    0
 [8,] -8.982907e-03 -1.274164e-03 -0.008974498    1
 [9,] -3.553964e-02 -1.274315e-03 -0.035719480    1
[10,] -3.549827e-02 -3.549068e-02 -0.035742288    1
[11,] -8.996946e-03 -3.540513e-02 -0.008974498    1
[12,] -3.093452e-04 -3.549068e-02 -0.035742288    1
[13,] -5.797193e-04 -3.505517e-04 -0.035742288    1
[14,] -3.577354e-02 -7.726913e-03 -0.035698152    1
[15,] -3.553951e-02 -2.650024e-05 -0.035719480    1
[16,] -5.813265e-04 -4.497953e-05 -0.035698152    1
```

**Comment by katzfuss** [comment]
The differences are definitely too large to be solely due to numerical error, but it is fairly difficult for me to debug like this. 

I would suggest to start with very simple examples, like n=3, m=2, M=1, where the MRA should be exact (Sigma = (UU')^{-1}), and then seeing exactly where things go wrong.

Note that we do want a sparse n x n matrix at the end, not a dense n x (m+1) matrix. There is no need to "convert back."

**Comment by marcinjurek** [comment]
Ok, so my U is now a square root of Sigma^{-1} but when I take the U generated by `U_NZentries()` and calculate UU' (also tried U'U) it is not equal to Sigma^{-1}... To be more precise, I'm looking at U_y returned by `U_NZentries`, or the rows and columns corresponding to the latent variables only. And it's all in this exact case (m=1, r=1, exponential covariance, 1d etc, where it should be exact).

**Comment by katzfuss** [comment]
I think you need to first compute Sigma.all=(UU')^{-1), and then extract the rows and columns corresponding to y from that matrix.

**Comment by marcinjurek** [comment]
Still trying to figure it out I'm reading [Shaefer (2017)](https://arxiv.org/pdf/1706.02205.pdf). In section 2.3, at the bottom of p. 9 it says:

__It is important to observe that the step min(i,j) is the first time when the entry \Theta_{i,j} is used in the Cholesky decomposition. Therefore, if we know that Cov[X_iX_j|{X_r}_{1<=r<=min(i,j)}] = 0, we need neither update nor even know the value of \Theta_{i,j}.__

I believe this is what you refered to in the overleaf document, and this is why calculation of U (U_Y, to be precise) only requires M=\Sigma \circ S, not the entire \Sigma.

However, it doesn't seem to work for me. Moreover, I don't understand why it should. 

My code currently can produce a U such that UU'=M^{-1}. But UU' != Sigma^{-1}. And it seems to make sense. If UU' = Sigma^{-1} then Sigma=M which is not true, b/c M=Sigma \circ S.

I also made sure that the old code works. The reason why it didn't yesterday was because of this problem with ordering (#37). Once I force the desired ordering U_{old} that the old code produces is indeed such that U_{old}(U_{old})' = Sigma^{-1}. However, U_{old} != U, and U_{old}(U_old)' != M^{-1}.

**Comment by marcinjurek** [comment]
Seems like this might be solved in #43 for the MRA case. Need to see how to easily pass the `NNarray`-shaped matrix of covariances to `U_NZentries()`
